### PR TITLE
support a new custom-registry-ca config option

### PR DESCRIPTION
**Pull request by kwmonroe**
Fixes: https://bugs.launchpad.net/layer-container-runtime-common/+bug/1831153

This PR adds a new `custom-registry-ca` config option which can be set to the base64-encoded CA cert of an external registry that uses untrusted certs.  This allows container runtimes (docker and containerd) to pull from those registries over https by installing the custom CA at the system level.

I switched the test methodology over to `tox` and added reactive handler unit tests.  To do this, i needed to rename the reactive py module to `s/-/_`, but I didn't notice any ill effects from this when upgrading an existing `containerd` to a local charm built with this PR.

Without this PR, if you tried to access an external self-signed registry from a containerd unit, you'd get:

```bash
containerd/0:~$ sudo wget --certificate=/root/cdk/server.crt --private-key=/root/cdk/server.key https://172.31.26.2:5000/v2
Connecting to 172.31.26.2:5000... connected.
ERROR: cannot verify 172.31.26.2's certificate, issued by ‘CN=54.67.83.243’:
  Unable to locally verify the issuer's authority.
To connect to 172.31.26.2 insecurely, use `--no-check-certificate'.
```

By setting this config option, the CA used to sign that registry's cert is installed into `/etc/ssl/certs` on the containerd units and the error is resolved:

```bash
juju config containerd custom-registry-ca=$(base64 ./ca.crt)
...

containerd/0:~$ sudo wget --certificate=/root/cdk/server.crt --private-key=/root/cdk/server.key https://172.31.26.2:5000/v2
Connecting to 172.31.26.2:5000... connected.
HTTP request sent, awaiting response... 200 OK
```

### CODE REVIEW

**Comment by kwmonroe** [comment]
I tested this with both subordinate runtimes on a k8s-worker against a docker-registry that had certs signed by a CA that was unknown to the k8s cluster:

### docker
From the `docker/0` unit without configuring this new option:

```bash
docker/0:~$ docker pull 172.31.26.2:5000/defaultbackend-amd64:1.5
Error response from daemon: Get https://172.31.26.2:5000/v2/: x509: certificate signed by unknown authority
```

Upgrade docker to a local charm built with this PR, configure the external CA, and try again:

```bash
juju config docker custom-registry-ca=$(base64 ./ca_dr.crt)
...

docker/0:~$ docker pull 172.31.26.2:5000/defaultbackend-amd64:1.5
1.5: Pulling from defaultbackend-amd64
65f4220de95d: Pull complete
Digest: sha256:4dc5e07c8ca4e23bddb3153737d7b8c556e5fb2f29c4558b7cd6e6df99c512c7
Status: Downloaded newer image for 172.31.26.2:5000/defaultbackend-amd64:1.5
172.31.26.2:5000/defaultbackend-amd64:1.5
```

### containerd
From the `containerd/0` unit without configuring this new option:

```bash
containerd/0:~$ sudo ctr image pull 172.31.26.2:5000/defaultbackend-amd64:1.5
ctr: failed to resolve reference "172.31.26.2:5000/defaultbackend-amd64:1.5": failed to do request: Head https://172.31.26.2:5000/v2/defaultbackend-amd64/manifests/1.5: x509: certificate signed by unknown authority
```

Upgrade containerd to a local charm built with this PR, configure the external CA, and try again:

```bash
juju config containerd custom-registry-ca=$(base64 ./ca_dr.crt)
...

containerd/0:~$ sudo ctr image pull 172.31.26.2:5000/defaultbackend-amd64:1.5
172.31.26.2:5000/defaultbackend-amd64:1.5:                                        resolved       |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:4dc5e07c8ca4e23bddb3153737d7b8c556e5fb2f29c4558b7cd6e6df99c512c7: done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:65f4220de95d2e3d12484679abe7bb33323b1fd3ef681d878f1d2bc5abc8ee06:    done           |++++++++++++++++++++++++++++++++++++++|
config-sha256:b5af743e598496e8ebd7a6eb3fea76a6464041581520d1c2315c95f993287303:   done           |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.1 s                                                                    total:   0.0 B (0.0 B/s)
unpacking linux/amd64 sha256:4dc5e07c8ca4e23bddb3153737d7b8c556e5fb2f29c4558b7cd6e6df99c512c7...
done
```

File: reactive/container_runtime_common.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        except Exception:
+            status.blocked('Invalid base64 value for custom-registry-ca config')
+            return
+        else:
+            host.install_ca_cert(_ca, name='juju-custom-registry')
+            charm = hookenv.charm_name()
+            hookenv.log('Custom registry CA has been installed for {}'.format(charm))
+
+            # NB: containerd auto restarts when config changes. docker does not,
+            # so manage the appropriate layer flags to recycle dockerd.
```
Comment (Cynerva) [inline reactive/container_runtime_common.py]: Is there a race condition with containerd? I'm worried that containerd's [config_changed](https://github.com/charmed-kubernetes/charm-containerd/blob/a77f8a2b26b93fdaa2af9d5fd0554fed9dd4232e/reactive/containerd.py#L328) handler might be called before install_custom_ca, not after.

**Comment by kwmonroe** [comment]
Added a restart handler to `charm-containerd` to address previous concern.  This PR is now dependent on https://github.com/charmed-kubernetes/charm-containerd/pull/37.

I've retested this + the containerd PR and in all cases (a few dozen), I see `restart_containerd` being queued/handled after `install_custom_ca`:

```
unit-containerd-0: 13:35:37 INFO unit.containerd/0.juju-log Invoking reactive handler: reactive/container_runtime_common.py:31:install_custom_ca
unit-containerd-0: 13:35:37 INFO unit.containerd/0.juju-log Installing new CA cert at: /usr/local/share/ca-certificates/juju-custom-registry.crt
...
unit-containerd-0: 13:35:38 INFO unit.containerd/0.juju-log Custom registry CA has been installed for containerd
unit-containerd-0: 13:35:38 DEBUG unit.containerd/0.juju-log tracer>
tracer: set flag containerd.restart
tracer: ++   queue handler reactive/containerd.py:421:restart_containerd
...
unit-containerd-0: 13:35:38 INFO unit.containerd/0.juju-log Invoking reactive handler: reactive/containerd.py:421:restart_containerd
unit-containerd-0: 13:35:38 INFO unit.containerd/0.juju-log status-set: maintenance: Restarting containerd
```

**Comment by Cynerva** [comment]
Thanks for the follow up. LGTM :+1:

**Pull request merged**